### ISSUE TITLE
Missing faces on Nvidia Curie (Geforce 6000/7000) running with Nvidia 304 driver and ATI R300 ones (regression)

### ISSUE THREAD

**Issue by illwieckz** [DaemonEngine/Daemon#376]
Only Nvidia Curie (Geforce 6000/7000) running with Nvidia 304 driver are affected, that's a regression.

This bug is a combination of two:

- #375 _Bad normal map channel detection, seems to be related to missing ARB_texture_compression_rgtc_
- a mistake introduced in #356 _Do not build GLSL shaders for disabled features_ especially in c5c18ce _do not build unused GLSL shader macro permutation_.

The bug makes face disappear or being badly rendered.

Unvanquished 0.51:

[![wrong normal format detection disables shader](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-29_011631_000.png)](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-29_011631_000.png)

[![wrong normal format detection disables shader](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-29_011648_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-29_011648_000.jpg)

Unvanquished 0.52:

[![wrong normal format detection disables shader](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-29_010422_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-29_010422_000.jpg)

[![wrong normal format detection disables shader](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-29_010437_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-29_010437_000.jpg)

[![wrong normal format detection disables shader](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-29_010504_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-29_010504_000.jpg)

[![wrong normal format detection disables shader](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-29_010527_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-29_010527_000.jpg)

[![wrong normal format detection disables shader](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-29_011817_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-29_011817_000.jpg)

**Comment by illwieckz** [comment]
The bug is caused by this:

https://github.com/DaemonEngine/Daemon/blob/f1b25a82fd7ad879a3e0ae80e300204d9460d5e2/src/engine/renderer/gl_shader.cpp#L1085-L1089

If a normal map is detected to embed an height map in alpha channel (wrongly or not, see #375 for a case of wrong detection), and at the same time both normal and relief mapping are disabled, the whole shader permutation is disabled because it has to be disabled (both normal and relief mapping are disabled so there is no need for those operations), but that will be the shader permutation that will be used because of the unused but detected height map in alpha channel.

The bug was verified on Nvidia Geforce 6150LE, 6600 Silencer, 7500LE using Nvidia 304 drivers.

An easy fix is to clear the shader boolean telling the normal map embeds an height map if there is no possible usage of that normal map file at all (because both normal and relief mapping are disabled).

Here is the result of such fix on the Geforce 6600 Silencer (it was also verified to fix the issue on other affected cards):

[![wrong normal format detection disables shader](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-30_234455_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-30_234455_000.jpg)

[![wrong normal format detection disables shader](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-30_234509_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-30_234509_000.jpg)

[![wrong normal format detection disables shader](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-30_234517_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-30_234517_000.jpg)

[![wrong normal format detection disables shader](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-30_234539_000.jpg)](https://dl.illwieckz.net/b/daemon/bugs/wrong-normal-format-detection-disables-shader/unvanquished_2020-09-30_234539_000.jpg)

**Comment by illwieckz** [comment]
It's possible to reproduce the bug on any Mesa driver this way:

```
MESA_EXTENSION_OVERRIDE=-GL_ARB_texture_compression_rgtc ./daemon
```

given the engine is compiled from `0.52.0/sync` branch and `lowest` preset is applied.

**Comment by illwieckz** [comment]
It is confirmed ATI R300 cards are affected too, with exact same artifacts.

Also, modern AMD GPU (GCN) with explicitely disabled RGTC extension on Mesa side was displaying the exact same artifacts, making the bug easy to reproduce even without access to the hardware. :heart: Mesa.

PR #377 is known to fix the issue on all known affected hardware.

**Comment by illwieckz** [comment]
Fixed twice by #385 and #377.
